src/main/java/org/truffleruby/language/globals/GlobalVariableStorage.java

```diff
@@ -26,26 +29,44 @@
 
     private volatile Object value;
 
+    private final Object defaultValue;
     private final DynamicObject getter;
     private final DynamicObject setter;
+    private final DynamicObject isDefined;
+
+    GlobalVariableStorage(Object defaultValue, DynamicObject getter, DynamicObject setter, DynamicObject isDefined) {
+        this.defaultValue = defaultValue;
+        this.value = UNSET_VALUE;
+        this.getter = getter;
+        this.setter = setter;
+        this.isDefined = isDefined;
+        assert ((getter == null) == (setter == null)) & ((getter == null) == (isDefined == null));
+    }
 
-    GlobalVariableStorage(Object value, DynamicObject getter, DynamicObject setter) {
+    GlobalVariableStorage(Object value, Object defaultValue, DynamicObject getter, DynamicObject setter, DynamicObject isDefined) {
+        this.defaultValue = defaultValue;
         this.value = value;
         this.getter = getter;
         this.setter = setter;
-        assert (getter != null) == (setter != null);
+        this.isDefined = isDefined;
+        assert ((getter == null) == (setter == null)) & ((getter == null) == (isDefined == null));
     }
```

Comment:
I think we might be able to simplify these constructors a bit, but it can be done later after the other PRs for global variables.